kernel/src/table_changes/physical_to_logical.rs

```diff
@@ -0,0 +1,191 @@
+use crate::{DeltaResult, Engine, Error, Expression, ExpressionRef, FileMeta};
+
+use super::resolve_dvs::ResolvedCdfScanFile;
+use super::scan_file::{CdfScanFile, CdfScanFileType};
+
+/// Returns a map from change data feed column name to an expression that generates the row data.
+#[allow(unused)]
+fn get_generated_columns(scan_file: &CdfScanFile) -> DeltaResult<HashMap<&str, Expression>> {
+    let timestamp = Scalar::timestamp_from_millis(scan_file.commit_timestamp)?;
+    let version = scan_file.commit_version;
+    let change_type: Expression = match scan_file.scan_type {
+        CdfScanFileType::Cdc => column_expr!("_change_type"),
+        CdfScanFileType::Add => "insert".into(),
+
+        CdfScanFileType::Remove => "delete".into(),
+    };
+    let expressions = [
+        ("_change_type", change_type),
+        ("_commit_version", Expression::literal(version)),
+        ("_commit_timestamp", timestamp.into()),
+    ];
+    Ok(expressions.into_iter().collect())
+}
+
+/// Generates the expression used to convert physical data from the `scan_file` path into logical
+/// data matching the `global_state.logical_schema`
+#[allow(unused)]
+fn get_expression(
+    scan_file: &CdfScanFile,
+    logical_schema: &StructType,
+    all_fields: &[ColumnType],
+) -> DeltaResult<Expression> {
+    let mut generated_columns = get_generated_columns(scan_file)?;
+    let all_fields = all_fields
+        .iter()
+        .map(|field| match field {
+            ColumnType::Partition(field_idx) => {
+                let field = logical_schema.fields.get_index(*field_idx);
+                let Some((_, field)) = field else {
+                    return Err(Error::generic(
+                        "logical schema did not contain expected field, can't transform data",
+                    ));
+                };
+                let name = field.physical_name();
+                let value_expression =
+                    parse_partition_value(scan_file.partition_values.get(name), field.data_type())?;
+                Ok(value_expression.into())
+            }
+            ColumnType::Selected(field_name) => {
+                // Remove to take ownership
+                let generated_column = generated_columns.remove(field_name.as_str());
+                Ok(generated_column.unwrap_or_else(|| ColumnName::new([field_name]).into()))
+            }
+        })
+        .try_collect()?;
+    Ok(Expression::Struct(all_fields))
+}
+
+/// Gets the physical schema that will be used to read data in the `scan_file` path.
+#[allow(unused)]
+fn get_read_schema(scan_file: &CdfScanFile, read_schema: &StructType) -> SchemaRef {
+    if scan_file.scan_type == CdfScanFileType::Cdc {
+        let change_type = StructField::new("_change_type", DataType::STRING, false);
+        let fields = read_schema.fields().cloned().chain(iter::once(change_type));
+        StructType::new(fields).into()
+    } else {
+        read_schema.clone().into()
+    }
```

Comment:
does this mean we are ignoring `_change_type` if it's written for Adds/removes? protocol says:
> In a table with Change Data Feed enabled, the data Parquet files referenced by add and remove actions are allowed to contain an extra column _change_type.

note it does say to ignore, but i think that means for normal reads (not computing CDF):
> This column is not present in the table's schema. When accessing these files, readers should disregard this column and only process columns defined within the table's schema.

Reply:
hmm good point. I haven't seen any tables with `_change_type` in the parquet file.  Also here isn't an easy way to use `_change_type` if present, and have the default we compute 🤔 

Tho it seems like we can safely ignore `_change_type` in the parquet file because delta spark uses `_change_type = insert` [for the entire file](https://github.com/delta-io/delta/blob/master/spark/src/main/scala/org/apache/spark/sql/delta/files/CdcAddFileIndex.scala#L69).

Reply:
good find! that sounds good - perhaps we can document those protocol quotes somewhere as well as the delta-spark link? just since this seems slightly at odds with the protocol? and to be clear i agree with the decision but this we should document all these findings :)